CORE-7996 Trial: build in an evaluation period license

pgellert opened the pull request:
Create a builtin fallback license that allows customers to trial
Enterprise features for a trial period starting at cluster creation.

Fixes https://redpandadata.atlassian.net/browse/CORE-7996

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes
### Features

* After the cluster is first formed, a trial license is automatically loaded to provide an evaluation period of enterprise features.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert commented:
/ci-repeat

oleiman commented on src/v/features/feature_table.cc:
```diff
@@ -297,6 +300,30 @@ class feature_table::probe {
+    auto expiry = std::chrono::duration_cast<std::chrono::seconds>(
+      expiry_time.time_since_epoch());
+
+    if (std::getenv("__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE") != nullptr) {
+        // For testing, use an expired trial license
+        expiry = 0s;
+    }
```
thought (pedantic): consider two cases 
1. no builtin trial license present (i.e. `nullopt`) 
2. builtin trial license expired

Intuitively these seem distinct, but I'm not sure there's a meaningful difference from a testing perspective. wdyt?

pgellert replied:
I've changed the (so far unused) `revoke_license()` method to also revoke the trial license and added some unit tests around this to demonstrate the behaviour of `should_sanction`. I suspect that `__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE` will be enough for ducktape tests (license nag, upgrade blocking tests, etc.), but we can tweak this env var in the future if necessary.

oleiman commented on src/v/features/feature_table.cc:
```diff
@@ -724,8 +727,19 @@ void feature_table::set_builtin_trial_license(
+          = _bootstrap_builtin_trial_license;
+        return ret;
+    } else {
+        return _license; // std::nullopt
```
nitpick: I'd just return `std::nullopt` explicitly (esp since you've written it out anyway 😛)

pgellert replied:
I chose this because I did not want to declare another static variable since I am returning a const& here. But yeah I could just do that instead.

oleiman replied:
oh, i see it. i didn't parse the return type properly. yeah that's weird. no worries.

oleiman commented on src/v/features/feature_table.cc:
```diff
@@ -724,8 +727,19 @@ void feature_table::set_builtin_trial_license(
-const std::optional<security::license>& feature_table::get_license() const {
-    return _license ? _license : _builtin_trial_license;
+const std::optional<security::license>&
+feature_table::get_license(bool use_bootstrap_license) const {
```
thought: there are a lot of `use_bootstrap_license = false` cases, so it could make sense to add a new accessor e.g. `feature_table::get_license_or_bootstrap`. 

less code motion in that case, but the compiler isn't much help.

pgellert replied:
I went with the even more specific `should_sanction` now

michael-redpanda commented on src/v/features/feature_table.cc:
```diff
@@ -686,10 +716,30 @@ void feature_table::set_license(security::license license) {
 
+void feature_table::set_builtin_trial_license(
+  model::timestamp cluster_creation_timestamp) {
+    if (ss::this_shard_id() == 0) {
+        vlog(featureslog.debug, "Initializing builtin trial license.");
+    }
```
Is this conditional just for the log message?

pgellert replied:
Yes, to only log on shard 0 and not all cores

michael-redpanda replied:
Maybe log the timestamp as well

michael-redpanda commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2407,7 +2407,7 @@ void admin_server::register_features_routes() {
           res.loaded = false;
           const auto& ft = _controller->get_feature_table().local();
-          const auto& license = ft.get_license();
+          const auto& license = ft.get_license(false);
```
we do want to return the evaluation license via the `get_license` endpoint

oleiman replied:
sort of related to this - the evaluation license has to supersede all functional effects of license enforcement, but

1. should the response from `GET /v1/features/enterprise` look like there's a real license installed? or 
2. should we suppress the license nag during the trial period?
3. should the `has_valid_license` bool in phone-home telemetry be an enum? or maybe we add an additional field for "evaluation_period"?

pgellert replied:
IMO yes, the evaluation period should work as if a real trial license had been installed (i.e., don't nag until expiry, show that there is a trial license installed, expiry time metric > 0).

```
% rpk cluster license info
LICENSE INFORMATION
===================
License status:     valid
License violation:  false
Organization:       Redpanda Built-In Evaluation Period
Type:               free_trial
Expires:            Dec 6 2024
```

> should the has_valid_license bool in phone-home telemetry be an enum? or maybe we add an additional field for "evaluation_period"?

Probably easiest to add a new field for "evaluation_period". I guess this could be derived from the license hash (`id_hash == 0`) but probably best to add it explicitly.

oleiman replied:
Aha, totally missed the free trial  type on the license!

michael-redpanda replied:
The request from product is to have it return "trial" just like if it's a generated trial license.  We should be able to tell that it's the evaluation by:
* The organization, and
* The empty SHA256 checksum

michael-redpanda commented on src/v/features/feature_table.h:
```diff
@@ -720,6 +735,14 @@ class feature_table {
     // Currently loaded redpanda license details
     std::optional<security::license> _license;
 
+    // Built in trial license to fall back to if there is no license set
+    std::optional<security::license> _builtin_trial_license;
+
+    // On startup the `_builtin_trial_license` is not available until the
+    // cluster is created. Until then, we fall back to this permissive
+    // bootstrap license.
+    security::license _bootstrap_builtin_trial_license;
+
```
I wonder if there's a way to only need to hold one copy of the license.  Would need to ensure that the built-in being set via `set_builtin_trial_license` can't override the true license.

Would also maybe need a flag indicating whether or not to return a license from the `get_license` endpoint if it's the built in eval vs the other two.

pgellert replied:
I've added the `should_sanction` method now which avoids the complexity of `_bootstrap_builtin_trial_license `

dotnwat commented on src/v/cluster/metrics_reporter.cc:
```diff
@@ -312,10 +313,14 @@ ss::future<> metrics_reporter::try_initialize_cluster_info() {
 
+    _cluster_info.creation_timestamp = first_cfg.header().first_timestamp;
+    co_await _feature_table.invoke_on_all([&](features::feature_table& ft) {
+        ft.set_builtin_trial_license(_cluster_info.creation_timestamp);
```
what do we do if creation timestamp is bogus (e.g. ntp isn't running and clocks aren't accurate)?

pgellert replied:
Yeah, that's a good point. Specifically, when only the creation timestamp is bogus, but now the time is accurate, right? Surely, all bets are off if the time is bogus throughout the cluster's runtime (e.g., license checks against the timestamp in the license would break equally). Do you have any ideas? The only improvement I can think of is sanity-checking that the creation timestamp is in the past, but it's hard to determine the cluster creation time retrospectively because the prefix of the controller log is snapshotted away after ~60s.

BenPope commented on src/v/features/feature_table.cc:
```diff
@@ -724,8 +727,19 @@ void feature_table::set_builtin_trial_license(
+    if (_license) {
+        return _license;
+    } else if (_builtin_trial_license) {
+        return _builtin_trial_license;
```
Is it possible to not set this? `revoke_license` doesn't unset this.

My goal is to test behaviour on license expiry.

BenPope replied:
Ahh: `__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE`

pgellert replied:
Yeah `__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE ` can be used for ducktape tests, `revoke_license` for unit/fixture tests.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/46c4f7959b8f09a03e3acb5f21208ea0079cca70..72f67bc369ce94c94284a28d78997b7463b8b59e):
* Introduced a `should_sanction` method that returns `false` until the cluster birth-based trial license is set. This avoids the need for the node-start-based cluster license and its ignoring parameter in `get_license`.
* Renamed the trial license org to "Redpanda Built-In Evaluation Period"

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/72f67bc369ce94c94284a28d78997b7463b8b59e..c3eb6f95e08e0bb592835eda87ab079e63123f0f):
* fix a bug (negated boolean ahead of `should_sanction`)
* add a unit test to demonstrate the expected `feature_table::should_sanction` behaviour
* have `revoke_license` revoke the builtin license as well

pgellert commented:
/ci-repeat

michael-redpanda commented on src/v/features/feature_table.h:
```diff
@@ -735,6 +735,10 @@ class feature_table {
     // Built in trial license to fall back to if there is no license set
     std::optional<security::license> _builtin_trial_license;
 
+    // Whether _builtin_trial_license has ever been initialized
+    // Used for implementing revoking the trial license for testing
+    bool _builtin_trial_license_initialized{false};
```
ahh thanks for the comment... was questioning it for a second

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--a46eb508-7020-484d-95a5-294e8fba4a30-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/57127#0192bf96-5477-437e-b9d1-cef05cec4b52:
```
"rptest.tests.rbac_test.RBACLicenseTest.test_license_nag"
"rptest.tests.redpanda_oauth_test.OIDCLicenseTest.test_license_nag.authn_config=.sasl_mechanisms.OAUTHBEARER.SCRAM"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/57127-->
<!--30d4f3fd-efe3-40cd-9504-930c4d901782-->
<!--title-->
## Retry command for [Build#57127](https://buildkite.com/redpanda/redpanda/builds/57127)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_license_violation@{"feature":3,"with_license":false}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_license_violation@{"feature":2,"with_license":false}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_license_violation@{"feature":10,"with_license":false}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_license_violation@{"feature":6,"with_license":false}
tests/rptest/tests/redpanda_kerberos_test.py::RedpandaKerberosLicenseTest.test_license_nag
```
<!--/body-->

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--f26105de-b6e0-49ca-aa83-bcbe83c034e4-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/57127#0192bf52-23a0-4741-ba44-a44402e57758 have failed and will be retried
```
cluster_cloud_metadata_rpfixture
gtest_cluster_cloud_metadata_rpfixture
```
<!--/body-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--da363a21-bafe-4f58-81e6-adeb848af61f-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/57127#0192bf96-5477-437e-b9d1-cef05cec4b52
<!--/body-->

IoannisRP commented on src/v/features/feature_table.cc:
```diff
@@ -686,10 +715,35 @@ void feature_table::set_license(security::license license) {
+}
+
+bool feature_table::should_sanction() const {
+    if (_license) {
+        return _license->is_expired();
+    } else if (_builtin_trial_license) {
+        return _builtin_trial_license->is_expired();
+    }
+
+    // While we are yet to initialize _builtin_trial_license on cluster
+    // creation, be permissive
+    return _builtin_trial_license_initialized;
```
```suggestion
    auto license = get_license();
    if (!license) {
        // ...
        return _builtin_trial_license_initialized;
    } 
    return license->is_expired();    
```
are these 2 identical? Or am i missing an edge case?

pgellert replied:
Yes, they are. I'm happy to switch to that if you think that's simpler.

IoannisRP commented on src/v/features/feature_table.cc:
```diff
@@ -686,10 +715,35 @@ void feature_table::set_license(security::license license) {
+    if (ss::this_shard_id() == 0) {
+        vlog(featureslog.debug, "Initializing builtin trial license.");
+    }
+    _builtin_trial_license_initialized = true;
+    _builtin_trial_license = make_builtin_trial_license(
+      model::to_time_point(cluster_creation_timestamp));
```
nit: while `make_builtin_trial_license` doesn't seem to be able to throw, we should probably set the bool after it is initialized, mainly for sanity reasons.

pgellert replied:
Fixed

oleiman commented:
FYI - tripped over https://github.com/redpanda-data/redpanda/blob/fb388a0fcc98ba029e2d33b36b13ad19cc6334ac/tests/rptest/services/redpanda.py#L5358 while working on DT for the config stuff

That check is no longer meaningful w/ a trial license. could be

```python
            if license is None or not license['loaded'] or \
                        license['license']['type'] == 'free_trial':
```
or something like that.

BenPope reviewed (commented):
Looking good.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/c3eb6f95e08e0bb592835eda87ab079e63123f0f..7702ed18fa46ff2d852b74881a2da2db9abe41b9):
 * fixed the unit/fixture/dt test failures detected in CI
 * addressed code review feedback
 * introduced `get_configured_license` which never returns the evaluation period license (used for taking controller snapshots and for cloud storage-based whole cluster recovery)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/57223-->
<!--d5038433-d5d2-4cbf-ae4b-41ee907d3aa6-->
<!--title-->
## Retry command for [Build#57223](https://buildkite.com/redpanda/redpanda/builds/57223)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group
```
<!--/body-->

michael-redpanda commented on tests/rptest/tests/enterprise_features_license_test.py:
```diff
@@ -76,42 +76,51 @@ def _put_license(self):
 
         return result
 
+    def _disable_evaluation_period(self):
+        self.redpanda.set_environment(
+            {'__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE': True})
+        self.redpanda.restart_nodes(self.redpanda.nodes)
```
suggestion: maybe move this to `redpanda.py`, but this isn't a blocker

pgellert replied:
FWIW the restart is only needed if the brokers have already been started. Tests that manually start the brokers (empty `setUp` method, eg. the tests in `license_enforcement_test.py`) can set the env var before the first broker is started and the brokers will pick it up on startup.

In this test case, I chose to restart the nodes instead of switching to manually start them because that seemed easier.

oleiman replied:
this is all getting ripped up in my PR anyway. i think "get the test passing and move on" is the right course of action.

michael-redpanda commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2291,7 +2291,7 @@ admin_server::get_enterprise_handler(std::unique_ptr<ss::http::request>) {
 
     const auto& license
-      = _controller->get_feature_table().local().get_license();
+      = _controller->get_feature_table().local().get_configured_license();
```
Ok I see you revert these changes in later commits.  I'll resolve them

michael-redpanda reviewed (approved):
Just a few non-blocking comments, looks good

pgellert commented on tests/rptest/services/admin.py:
```diff
@@ -904,6 +904,25 @@ def get_license(self, node=None, timeout=None):
+
+        return resp is not None and resp.get('license', None) is not None \
+                and resp['license']['type'] == 'enterprise' \
+                and resp['license']['org'] == 'redpanda-testing'
```
See the comment on line 914. Maybe I should move it below the examples

michael-redpanda replied:
ahhhhhh gotcha

pgellert commented on tests/rptest/tests/license_enforcement_test.py:
```diff
@@ -46,8 +46,10 @@ def setUp(self):
     def test_license_enforcement(self, clean_node_before_recovery,
                                  clean_node_after_recovery):
-        installer = self.redpanda._installer
+        self.redpanda.set_environment(
+            {'__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE': True})
```
I didn't add it in the constructor because I only wanted it to apply to the first two tests but not `test_enterprise_cluster_bootstrap`. `start` and `restart` don't currently take extra environment variables; only extra configs (at the moment).

I guess I could factor out the first two tests into a separate test class, and then I could pass it into the constructor.

oleiman commented on tests/rptest/tests/cluster_features_test.py:
```diff
@@ -229,11 +229,13 @@ def test_license_upload_and_query(self):
+        def obtain_configured_license():
             lic = self.admin.get_license()
-            return (lic is not None and lic['loaded'] is True, lic)
+            return (self.admin.is_sample_license(lic), lic)
```
shouldn't this be `not sample_license`?

pgellert replied:
No, we're waiting to observe the sample license (the one available in CI) inserted on line 230. Maybe you thought `is_sample_license` returns true for the evaluation period license? Happy to change to a better name if you have suggestions. I went with this one because the sample license available in CI is retrieved with the `sample_license()` method.

oleiman replied:
Oops, yeah, naming confusion. No big deal

oleiman commented on tests/rptest/tests/enterprise_features_license_test.py:
```diff
@@ -76,42 +76,51 @@ def _put_license(self):
-    ])
-    def test_get_enterprise(self, with_license):
+    @matrix(with_license=[True, False], with_evaluation_period=[True, False])
+    def test_get_enterprise(self, with_license, with_evaluation_period):
```
might consider just marking these ok to fail temporarily and otherwise leaving exactly as is. your changes look fine but it'll be a pain to integrate with https://github.com/redpanda-data/redpanda/pull/23899

dotnwat commented on src/v/features/feature_table.cc:
```diff
@@ -686,12 +714,47 @@ void feature_table::set_license(security::license license) {
 }
 
+bool feature_table::should_sanction() const {
+    if (_license) {
+        return _license->is_expired();
+    } else if (_builtin_trial_license) {
+        return _builtin_trial_license->is_expired();
+    }
```
is it the case that you are creating a special carve-out for the trial license? it would seem that an alternative would have been to generate an actual license that would expire as expected, and re-use all of the existing machinery. is that because you can't generate a license without the private key? if so, would it have been an option to expand the semantics of the existing license abstraction as opposed to expressing a trial license explicitly?

pgellert replied:
We are able to create a license in the broker without the private key because internally the license is stored unencrypted, only licenses passed in from outside through the admin API are signed and validated.

We store the real license and the eval license separately (ie. have the eval license carved out) because we need a way to tell them apart at some points of the code. For example, we only recover the user-configured license during whole cluster recovery but not the evaluation period license, and we only write the user-configured license into the controller snapshot. We could represent this information differently (eg. introduce `security::license_type::evaluation_period` or add a new `source_type` field to `security::license`), but storing it as a separate field seemed simplest to me because it guarantees by design that we can't overwrite a real license with an evaluation period license. Storing the license in memory instead of writing it into the controller log also makes it easy to change the behaviour of the trial license later if the product asks change (eg. change the evaluation period length).

(fyi I've drafted up the alternative you proposed; it would look something like this: https://github.com/redpanda-data/redpanda/pull/23945)

Let me know what you think with the above in mind. I'm open to changing it, potentially even to this alternative implementation, but the reasons above are why the evaluation period license is currently carved out into a separate field.

dotnwat replied:
thanks for the detailed response. i think what we are doing in this PR is ok. actually, it's great--not generalizing until we get a few examples (currently N=2) makes a lot of sense. but maybe you can see where it could break down in the future: product asks for a new kind of license, and then we are asked for carve-outs for different features, etc... pretty soon the carve-out approach becomes unwieldy.

pgellert replied:
yeah makes sense. I would expect that product asks would mostly just change the way the evaluation period behaves and we would not need to introduce a new trial license type. But if we do, we can certainly think about ways of simplifying things at N=3.

michael-redpanda commented on src/v/cluster/feature_manager.cc:
```diff
@@ -330,7 +330,7 @@ void feature_manager::verify_enterprise_license() {
     }
 
-    const auto& license = _feature_table.local().get_configured_license();
+    const auto& license = _feature_table.local().get_license();
```
Yes confirmed this is good

pgellert commented:
/cdt

pgellert commented:
[CDT test](https://ci-artifacts.dev.vectorized.cloud/redpanda/57294/0192d8df-f019-4390-967d-8385bb1ab197/vbuild/ducktape/results/2024-10-29--001/report.html) failures:
 * IcebergRESTCatalogSmokeTest.test_basic -- https://redpandadata.atlassian.net/browse/CORE-8097
 * LargeMessagesTest.test_large_messages_throughput -- https://redpandadata.atlassian.net/browse/DEVPROD-2066
 * TestReadReplicaService.test_writes_forbidden -- https://redpandadata.atlassian.net/browse/CORE-7867
 * NodeWiseRecoveryTest.test_node_wise_recovery -- https://redpandadata.atlassian.net/browse/CORE-7977
 * PartitionBalancerTest.test_transfer_controller_leadership -- https://redpandadata.atlassian.net/browse/CORE-7799 (different test, same symptom as this one and a few others)
 * AlterConfigMixedNodeTest.test_alter_config_shadow_indexing_mixed_node -- seems unrelated, maybe this could be flaky on line 409 reaching `redpanda.remote.readreplica` instead of read?
 * DataMigrationsApiTest.* -- https://redpandadata.atlassian.net/browse/CORE-8064
 * DescribeTopicsTest.test_describe_topics_with_documentation_and_types -- seems unrelated (looks like the test needs updating with the new topic property `topic_property_iceberg_translation_interval_ms`)

emaxerrno commented on src/v/features/feature_table.cc:
```diff
@@ -297,6 +301,30 @@ class feature_table::probe {
+    auto expiry = std::chrono::duration_cast<std::chrono::seconds>(
+      expiry_time.time_since_epoch());
+
+    if (std::getenv("__REDPANDA_DISABLE_BUILTIN_TRIAL_LICENSE") != nullptr) {
```
the double prefix underscore is odd. we have so many notations we should unify. 

RPK has env variables for example as well that do not have this. @twmb - we should have a developer experience here unify these. the product flags feel disjointed.

michael-redpanda replied:
The intended use for this environmental variable is for testing and not for end-user consumption.  This pattern matches other environmental variables [here](https://github.com/redpanda-data/redpanda/blob/64375e957eecf4eff7ae4f63633c868046037854/src/v/features/feature_table.cc#L334), [here](https://github.com/redpanda-data/redpanda/blob/64375e957eecf4eff7ae4f63633c868046037854/src/v/features/feature_table.cc#L352)], and [here](https://github.com/redpanda-data/redpanda/blob/64375e957eecf4eff7ae4f63633c868046037854/src/v/features/feature_table.cc#L394)

emaxerrno replied:
i pinged @mattschumpert on this too to make sure all the vars are consistent. this should have product input for consistency. i only care about consistency, the prefix,postfix,dot-notation,etc doesn't matter as much to me as consistency across all product lines.

mattschumpert replied:
The ENV vars specifically are so internal facing and no user ever sees them unless their browsing the code that it only matters for redpanda devs I think. 

I've never once looked at any of these (only a couple of documented external ones).

the '__REDPANDA' seems to be our convention for internal, and seems consistent so it makes sense to me.

dotnwat replied:
Yeh, internal consumption only. Should never be in documentation.
